Flow/flow.h

```diff
@@ -0,0 +1,415 @@
+
+#ifndef CURRENT_FLOW_H
+#define CURRENT_FLOW_H
+
+#include "../port.h"
+
+#include <cassert>
+#include <functional>
+#include <iostream>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "lazy_instantiation.h"
+
+namespace flow {
+
+// Template logic to make dealing with LHS/RHS/VIA/E2E subflows more generic.
+enum class InputPolicy { DoesNotAccept = 0, Accepts = 1 };
+enum class OutputPolicy { DoesNotEmit = 0, Emits = 1 };
+
+template <InputPolicy, OutputPolicy>
+struct BeautifyDescription {};
+template <>
+struct BeautifyDescription<InputPolicy::DoesNotAccept, OutputPolicy::DoesNotEmit> {
+  static std::string run(const std::string& s) { return s; }
+};
+template <>
+struct BeautifyDescription<InputPolicy::Accepts, OutputPolicy::DoesNotEmit> {
+  static std::string run(const std::string& s) { return "... | " + s; }
+};
+template <>
+struct BeautifyDescription<InputPolicy::DoesNotAccept, OutputPolicy::Emits> {
+  static std::string run(const std::string& s) { return s + " | ..."; }
+};
+template <>
+struct BeautifyDescription<InputPolicy::Accepts, OutputPolicy::Emits> {
+  static std::string run(const std::string& s) { return "... | " + s + " | ..."; }
+};
+
+// Template logic to spawn message acceptors and message emitters.
+struct CanBeSpawnedToAccept {
+  virtual ~CanBeSpawnedToAccept() = default;
+  virtual std::function<void(int)> SpawnAcceptor(bricks::LazyInstantiationStrategy) const = 0;
+};
+
+struct CanBeSpawnedToEmit {
+  virtual ~CanBeSpawnedToEmit() = default;
+  virtual void SpawnEmitter(std::function<void(int)> f, bricks::LazyInstantiationStrategy) const = 0;
+};
+
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+struct SubflowRunner;  // {};
+
+template <>
+struct SubflowRunner<InputPolicy::DoesNotAccept, OutputPolicy::DoesNotEmit> {
+  virtual ~SubflowRunner() = default;
+};
+template <>
+struct SubflowRunner<InputPolicy::DoesNotAccept, OutputPolicy::Emits> : CanBeSpawnedToEmit {
+  virtual ~SubflowRunner() = default;
+};
+template <>
+struct SubflowRunner<InputPolicy::Accepts, OutputPolicy::DoesNotEmit> : CanBeSpawnedToAccept {
+  virtual ~SubflowRunner() = default;
+};
+template <>
+struct SubflowRunner<InputPolicy::Accepts, OutputPolicy::Emits> : CanBeSpawnedToAccept, CanBeSpawnedToEmit {
+  virtual ~SubflowRunner() = default;
+};
+
+// General fields for each subflow, complete or incomplete.
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+class Subflow : public SubflowRunner<INPUT, OUTPUT> {
+ public:
+  explicit Subflow(const std::string& description) : description_(description) {}
+
+  // No subflow should be left hanging.
+  virtual ~Subflow() {
+    if (still_hanging && !describe_called) {
+      // Too bad we can't throw an exception here. -- D.K.
+      // TODO(dkorolev): Unit-test this with a singleton policy with ostream and std::exit().
+      std::cerr << "Current leaked:\n" << Describe() << "\nCall `.Flow()`, etc., don't leave it hanging.\n";
```

Comment:
If the end user sees this error, it's already the full, top-level expression. I'd like to train people call those full expressions Currents! :-)